### DISCUSSION TITLE
C# - Upgrade from .sln to .slnx solution files

### DISCUSSION THREAD

**Discussion by Joy-less** [godotengine/godot-proposals#11389] [Scripting]
The `.sln` format for C# solution files (containing multiple projects) is a very [old and proprietary format](https://medium.com/c-sharp-programming/a-closer-look-at-visual-studios-new-slnx-file-format-521e15a20e96) from 1997.
Here's an example:
```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 17
VisualStudioVersion = 17.8.34322.80
MinimumVisualStudioVersion = 10.0.40219.1
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "RecordTypes", "RecordTypes.csproj", "{3322D5B8-4E43-42B6-BA18-6F6188ED8801}"
EndProject
Global
 GlobalSection(SolutionConfigurationPlatforms) = preSolution
  Debug|Any CPU = Debug|Any CPU
  Release|Any CPU = Release|Any CPU
 EndGlobalSection
 GlobalSection(ProjectConfigurationPlatforms) = postSolution
  {3322D5B8-4E43-42B6-BA18-6F6188ED8801}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
  {3322D5B8-4E43-42B6-BA18-6F6188ED8801}.Debug|Any CPU.Build.0 = Debug|Any CPU
  {3322D5B8-4E43-42B6-BA18-6F6188ED8801}.Release|Any CPU.ActiveCfg = Release|Any CPU
  {3322D5B8-4E43-42B6-BA18-6F6188ED8801}.Release|Any CPU.Build.0 = Release|Any CPU
 EndGlobalSection
 GlobalSection(SolutionProperties) = preSolution
  HideSolutionNode = FALSE
 EndGlobalSection
 GlobalSection(ExtensibilityGlobals) = postSolution
  SolutionGuid = {AE1350F1-0A7D-41AD-8016-79BE8B49D945}
 EndGlobalSection
EndGlobal
```
In April 2024, Microsoft released an alternative solution format called `.slnx`, which uses XML, like `.csproj` files.
```
<Solution>
  <Project Path="RecordTypes.csproj" />
</Solution>
```

It would be ideal for Godot to support the new `.slnx` format when it is rolled out. Currently, Godot appears to be [hardcoded to support only `.sln` files](https://github.com/godotengine/godot/blob/89001f91d21ebd08b59756841426f540b154b97d/modules/mono/editor/GodotTools/GodotTools.ProjectEditor/DotNetSolution.cs):
```py
if os.path.exists(f"{project_name}.sln"):
        for line in open(f"{project_name}.sln", "r", encoding="utf-8").read().splitlines():
            if line.startswith('Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}")'):
                proj_uuid = re.search(
                    r"\"{(\b[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-\b[0-9a-fA-F]{12}\b)}\"$",
                    line,
                ).group(1)
            elif line.strip().startswith("SolutionGuid ="):
                sln_uuid = re.search(
                    r"{(\b[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-\b[0-9a-fA-F]{12}\b)}", line
                ).group(1)
                break
```
```cs
private const string _projectDeclaration =
@"Project(""{{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}}"") = ""{0}"", ""{1}"", ""{{{2}}}""
EndProject";
```

If Godot switched to only support `.slnx` files, a lot of this proprietary parsing mess code could be removed and replaced with an XML parser.

Additionally, the developer experience when manually editing the `.sln` file will be improved, for example when renaming a project it is often necessary to modify the `.sln` file.

- Now supported by `msbuild`: https://github.com/dotnet/msbuild/issues/10266
- Support pending for `VSCode`: https://github.com/dotnet/vscode-csharp/issues/7048

Note that `.slnx` seems to be intended for `.NET 9.0`, so it might be best to wait a bit. Consider this a long-term discussion to be revisited in the future.

**Comment by raulsntos** [comment]
There is already an approved PR (queued for 4.6) to support `.slnx` solution files:

- https://github.com/godotengine/godot/pull/106031

After that, you'll be able to upgrade your `.sln` files to `.slnx` with the `dotnet sln migrate` command.

Using `.slnx` by default can't happen until most IDEs have decent support. It doesn't seem like VSCode is quite there yet, which would affect a big chunk of C# users (me included). And I'd rather wait for upstream templates to switch first (e.g.: `dotnet new console`).

Also, removing `.sln` support will likely never happen due to preserving compatibility. But the new  [Microsoft.VisualStudio.SolutionPersistence](https://www.nuget.org/packages/Microsoft.VisualStudio.SolutionPersistence) package implements all the parsing for both `.sln` and `.slnx` and we are already using it in the PR mentioned above, so there's no need to drop support for `.sln`.

**Reply by Herdo** [reply]
@raulsntos Most .NET IDEs (Visual Studio, VS Core, Rider) have support for `.slnx` nowadays. `dotnet new sln` also creates a `.slnx`.
